File: neetcode150/18_bit_manipulation/0268_missing_number.cpp
/*
 * =====================================================================
 * LeetCode 268 : Missing Number                                     (Easy)
 * https://leetcode.com/problems/missing-number/
 * Category   : Bit Manipulation
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given an array `nums` containing n DISTINCT numbers taken from
 *   0, 1, 2, ..., n, return the ONE number that is missing from the
 *   range. Example: [3,0,1] → 2. Must run in O(n) time and O(1)
 *   extra space.
 * ---------------------------------------------------------------------
 * INTUITION
 *   The array plus the missing value forms the full set {0..n}.
 *   Two algebraic lenses see the hole instantly:
 *     • SUM: the full range sums to n(n+1)/2; subtract the array's
 *       sum — the difference IS the missing number.
 *     • XOR: xor every index with every value. Every (index, value)
 *       pair that is present cancels (x ^ x = 0, exactly the 0136
 *       trick), leaving the index that never matched a value.
 * ---------------------------------------------------------------------
 * APPROACH 1 — XOR of all indices and values (BEST)
 *   acc = n; for i in 0..n-1: acc ^= i ^ nums[i]. Present pairs
 *   cancel; acc ends as the unmatched index = the missing number.
 *   No overflow possible — the sum approach can overflow int32 for
 *   large n; XOR cannot.
 *   Time  Complexity : O(n)   Space Complexity : O(1).
 * ---------------------------------------------------------------------
 * APPROACH 2 — Sum formula
 *   expected = n(n+1)/2; return expected − Σ nums. O(n) time and
 *   O(1) space, but n(n+1)/2 grows quadratically — an int32 can
 *   overflow for n beyond ~65536 without 64-bit arithmetic.
 *   Time  Complexity : O(n)   Space Complexity : O(1).
 * ---------------------------------------------------------------------
 * DEEP DIVE — why the XOR works, step by step
 *   We KNOW the full range {0..n} has n+1 distinct values. Its XOR
 *   depends only on n. Now xor in every ARRAY element:
 *       acc = n
 *       acc ^= 0 ^ nums[0] ^ 1 ^ nums[1] ^ ... ^ (n-1) ^ nums[n-1]
 *   Rearrange (commutative): every value v that IS present meets its
 *   own index v and cancels (v ^ v = 0). The one index m whose value
 *   is absent has nothing to cancel with — it is the sole survivor.
 *   Identical machinery to 0136 Single Number — same proof, new hat.
 * =====================================================================
 */

#include <iostream>      // std::cout — printing benchmark results
#include <vector>        // std::vector — the input array
#include <cstdint>       // std::int64_t — overflow-free sum lane
#include <climits>       // INT_MAX — scale check for the test table
#include <chrono>        // std::chrono::steady_clock — timing for benchmark
#include <sys/resource.h>// getrusage — peak RAM measurement for benchmark

/* =====================================================================
 * APPROACH 1 : XOR of all indices and values — pairs cancel
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Find the single value missing from 0..n in `nums`.
     * Inputs  : nums — n distinct integers from the range 0..n.
     * Output  : int — the missing value.
     */
    int missingNumber(const std::vector<int>& nums) {
        // Start with n: it is the only index with NO counterpart in
        // the array a priori (the array spans indices 0..n-1).
        int acc = static_cast<int>(nums.size());

        // Pair every index with the value living at it. Present pairs
        // cancel to 0; the missing index never finds its partner.
        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            acc ^= i ^ nums[i];
        }
        return acc;
    }
};

/* =====================================================================
 * APPROACH 2 : Sum formula — full range minus the array
 * ===================================================================== */
class Solution_2 {
public:
    /*
     * Purpose : Find the missing value using the arithmetic-series
     *           formula; the 64-bit lane dodges the quadratic growth.
     * Inputs  : nums — n distinct integers from the range 0..n.
     * Output  : int — the missing value.
     */
    int missingNumber(const std::vector<int>& nums) {
        // Full range {0..n} sums to n(n+1)/2 — compute in a 64-bit
        // lane because n(n+1) overflows int32 for n > 65535.
        std::int64_t n = static_cast<std::int64_t>(nums.size());
        std::int64_t expected = n * (n + 1) / 2;

        // Accumulate the array's actual sum — int64 again for safety.
        std::int64_t actual = 0;
        for (int v : nums) {
            actual += v;
        }

        // The difference is the missing value, and it always fits int.
        return static_cast<int>(expected - actual);
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches and compare time + peak RAM
 * ===================================================================== */
int main() {
    // Test table: { array, expected missing value } — the official
    // examples plus a large-range case where the sum grows.
    std::vector<std::pair<std::vector<int>, int>> tests = {
        {{3, 0, 1}, 2},
        {{0, 1}, 2},
        {{9, 6, 4, 2, 3, 5, 7, 0, 1}, 8},
        {{0, 1, 2, 3, 4, 5, 6, 7, 9}, 8}
    };

    // --- Approach 1: run every test, measure time + peak RAM ---
    struct rusage before_1, after_1;
    getrusage(RUSAGE_SELF, &before_1);
    auto start_1 = std::chrono::steady_clock::now();
    bool ok_1 = true;
    for (const auto& t : tests) {
        int got = Solution_1().missingNumber(t.first);
        bool pass = got == t.second;
        ok_1 = ok_1 && pass;
        std::cout << "  A1 missingNumber(" << t.first.size() << " elems) = "
                  << got << "  expected " << t.second
                  << (pass ? "  PASS" : "  FAIL") << "\n";
    }
    auto end_1 = std::chrono::steady_clock::now();
    getrusage(RUSAGE_SELF, &after_1);

    // nanoseconds → microseconds; ru_maxrss is ALREADY in KB on Linux.
    double us_1 = std::chrono::duration_cast<std::chrono::nanoseconds>(end_1 - start_1).count() / 1000.0;
    long mem_1 = after_1.ru_maxrss - before_1.ru_maxrss;  // KB

    // --- Approach 2: run every test, measure time + peak RAM ---
    struct rusage before_2, after_2;
    getrusage(RUSAGE_SELF, &before_2);
    auto start_2 = std::chrono::steady_clock::now();
    bool ok_2 = true;
    for (const auto& t : tests) {
        int got = Solution_2().missingNumber(t.first);
        bool pass = got == t.second;
        ok_2 = ok_2 && pass;
        std::cout << "  A2 missingNumber(" << t.first.size() << " elems) = "
                  << got << "  expected " << t.second
                  << (pass ? "  PASS" : "  FAIL") << "\n";
    }
    auto end_2 = std::chrono::steady_clock::now();
    getrusage(RUSAGE_SELF, &after_2);

    double us_2 = std::chrono::duration_cast<std::chrono::nanoseconds>(end_2 - start_2).count() / 1000.0;
    long mem_2 = after_2.ru_maxrss - before_2.ru_maxrss;  // KB

    // Summary lines: time in µs, memory delta in KB.
    std::cout << "Approach 1 (XOR pairs)       : time = " << us_1
              << " µs  peak-ram-delta = " << mem_1 << " KB\n";
    std::cout << "Approach 2 (sum formula)     : time = " << us_2
              << " µs  peak-ram-delta = " << mem_2 << " KB\n";

    std::cout << (ok_1 && ok_2
                  ? "PASS : both approaches find every missing number."
                  : "FAIL : at least one test mismatched.")
              << "\n";

    return 0;
}

File: neetcode150/18_bit_manipulation/0268_missing_number.py
"""
=====================================================================
LeetCode 268 : Missing Number                                     (Easy)
https://leetcode.com/problems/missing-number/
Category   : Bit Manipulation
---------------------------------------------------------------------
PROBLEM
    Given an array `nums` containing n DISTINCT numbers taken from
    0, 1, 2, ..., n, return the ONE number that is missing from the
    range. Example: [3,0,1] → 2. Must run in O(n) time and O(1)
    extra space.
---------------------------------------------------------------------
INTUITION
    The array plus the missing value forms the full set {0..n}. Two
    algebraic lenses see the hole instantly:
      • SUM: the full range sums to n(n+1)/2; subtract the array's
        sum — the difference IS the missing number.
      • XOR: xor every index with every value. Every present
        (index, value) pair cancels (x ^ x = 0, the 0136 trick),
        leaving the index that never found a matching value.
---------------------------------------------------------------------
APPROACH 1 — XOR of all indices and values (BEST)
    acc = n; for i: acc ^= i ^ nums[i]. Present pairs cancel; acc
    ends as the unmatched index = the missing number. No overflow.
    Time  Complexity : O(n)   Space Complexity : O(1)
    WHY it works: acc = n ^ 0 ^ nums[0] ^ 1 ^ nums[1] ^ ...
    Every present value v meets its own index v and cancels; the
    index m whose value is absent survives alone. Same machinery
    as 0136 Single Number — same proof, new hat.
---------------------------------------------------------------------
APPROACH 2 — Sum formula
    expected = n(n+1)/2; return expected − sum(nums).
    Time  Complexity : O(n)   Space Complexity : O(1)
    WARNING: n(n+1)/2 grows quadratically — in C++ it overflows
    int32 for n > 65535 (Python's unbounded ints dodge that, which
    is exactly why the XOR variant is the portable winner).
---------------------------------------------------------------------
REAL LIFE
    • Database consistency: detecting a missing sequential key (an
      auto-increment column with a deleted row) is sum-difference
      checking on a million-row table.
    • Network packet sequencing: missing packet numbers in a stream
      are found with the same range-difference identity.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import List, Tuple


# =====================================================================
# APPROACH 1 : XOR of all indices and values — pairs cancel (best)
# =====================================================================
class Solution_1:
    """
    Purpose : Find the single value missing from 0..n in `nums`.
    Inputs  : nums — n distinct integers from the range 0..n.
    Output  : int — the missing value.
    """

    def missingNumber(self, nums: List[int]) -> int:
        # Start with n: the only index with no counterpart a priori
        # (the array spans indices 0..n-1).
        acc = len(nums)

        # Pair every index with the value living at it. Present pairs
        # cancel to 0; the missing index never finds its partner.
        for i, num in enumerate(nums):
            acc ^= i ^ num
        return acc


# =====================================================================
# APPROACH 2 : Sum formula — full range minus the array
# =====================================================================
class Solution_2:
    """
    Purpose : Find the missing value using the arithmetic-series
              formula (n(n+1)/2 minus the array sum).
    Inputs  : nums — n distinct integers from the range 0..n.
    Output  : int — the missing value.
    """

    def missingNumber(self, nums: List[int]) -> int:
        # Full range {0..n} sums to n(n+1)/2. Python ints are
        # unbounded, so the quadratic growth never overflows here —
        # but note the C++ int32 equivalent would break at n > 65535.
        n = len(nums)
        expected = n * (n + 1) // 2

        # The difference between what SHOULD be there and what IS
        # there is exactly the missing value.
        return expected - sum(nums)


# =====================================================================
# BENCHMARK — run both approaches and compare time + peak memory
# =====================================================================
if __name__ == "__main__":
    # Test table: {array, expected missing value} — official examples
    # plus a large-range case where the sum grows quadratically.
    tests: List[Tuple[List[int], int]] = [
        ([3, 0, 1], 2),
        ([0, 1], 2),
        ([9, 6, 4, 2, 3, 5, 7, 0, 1], 8),
        (list(range(0, 100001))[:77777] + list(range(77778, 100001)), 77777),
    ]

    # --- Approach 1: time + peak memory (tracemalloc snapshot) ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok_1 = True
    for arr, exp in tests:
        got = Solution_1().missingNumber(arr)
        pass_ = got == exp
        ok_1 = ok_1 and pass_
        print(f"  A1 missingNumber({len(arr)} elems) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1 = time.perf_counter()
    _, peak_1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2: time + peak memory ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok_2 = True
    for arr, exp in tests:
        got = Solution_2().missingNumber(arr)
        pass_ = got == exp
        ok_2 = ok_2 and pass_
        print(f"  A2 missingNumber({len(arr)} elems) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1b = time.perf_counter()
    _, peak_2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # Summary lines: time in µs, peak memory in KB.
    print(f"Approach 1 (XOR pairs)       : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak_1 / 1024:.2f} KB")
    print(f"Approach 2 (sum formula)     : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak_2 / 1024:.2f} KB")

    print("PASS : both approaches find every missing number."
          if ok_1 and ok_2
          else "FAIL : at least one test mismatched.")
